cmake/modules/yaml.cmake

```diff
@@ -399,6 +441,57 @@ function(yaml_save)
   FILE(WRITE ${yaml_file} "${yaml_out}")
 endfunction()
 
+# Usage:
+#   yaml_generate(NAME <name> TARGET <tgt> [FILE <file>])
+#
+# Schedule writing the YAML context <name>, expanding generator expressions
+# referring to the specified target, to a file. The operation will not be
+# performed before this function returns; instead, the file will be written as
+# a pre-build step of the specified target.
+#
+# NAME <name>:  Name of the YAML context
+# TARGET <tgt>: Target to be used as reference for generator expansions.
+# FILE <file>:  Path to file to write the context.
+#               If not given, then the FILE property of the YAML context will be
+#               used. In case both FILE is omitted and FILE property is missing
+#               on the YAML context, then an error will be raised.
+#
+function(yaml_generate)
+  cmake_parse_arguments(ARG_YAML "" "TARGET;NAME;FILE" "" ${ARGN})
+
+  zephyr_check_arguments_required(${CMAKE_CURRENT_FUNCTION} ARG_YAML NAME TARGET)
+  internal_yaml_context_required(NAME ${ARG_YAML_NAME})
+
+  set(yaml_target ${ARG_YAML_TARGET})
+  get_target_property(yaml_file ${ARG_YAML_NAME}_scope FILE)
+  if(NOT yaml_file)
+    zephyr_check_arguments_required(${CMAKE_CURRENT_FUNCTION} ARG_YAML FILE)
+  endif()
+  if(DEFINED ARG_YAML_FILE)
+    set(yaml_file ${ARG_YAML_FILE})
+  endif()
+
+  get_property(json_content TARGET ${ARG_YAML_NAME}_scope PROPERTY JSON)
+
+  if(EXISTS ${yaml_file})
+    FILE(RENAME ${yaml_file} ${yaml_file}.bak)
+  endif()
+  FILE(GENERATE OUTPUT ${yaml_file}.tmp
+    CONTENT "${json_content}"
+    TARGET ${yaml_target}
+  )
+  add_custom_target(${ARG_YAML_NAME}_yaml_filter ALL DEPENDS ${yaml_file}.tmp ${yaml_target})
+  add_custom_command(TARGET ${yaml_target} PRE_BUILD
```

Comment:
that you could want to save known data as soon as you have them.
For example, sysbuild could want to save as much info as possible **before** CMake config step of individual images are executed (and potentially breaks the configure step).

This would provide knowledge to users / IDEs / other tools about configuration files, included images, etc. even if one of those fails.